### PR TITLE
WIP: runtime support for static virtual interface methods

### PR DESCRIPTION
**Pull request by trylek**
I have refactored my earlier changes based on the principle that the virtual static methods shouldn't go in the vtable. I'm now at the point of runtime lookup required by getCallInfo.

Thanks

Tomas

### CODE REVIEW

File: src/coreclr/vm/methodtablebuilder.cpp [hunk 1]
```diff
@@ -5021,14 +5029,14 @@ MethodTableBuilder::ValidateMethods()
...
         if (it.IsMethodImpl())
         {
-            if (!IsMdVirtual(it.Attrs()))
+            if (!IsMdVirtual(it.Attrs()) && !IsMdStatic(it.Attrs()))
             {   // Non-virtual methods cannot participate in a methodImpl pair.
                 BuildMethodTableThrowException(IDS_CLASSLOAD_MI_MUSTBEVIRTUAL, it.Token());
             }
         }
 
         // Virtual static methods are not allowed.
```
Comment (davidwrighton) [inline src/coreclr/vm/methodtablebuilder.cpp:5040]: Adjust comment

File: src/coreclr/vm/methodtablebuilder.cpp [hunk 2]
```diff
@@ -1180,17 +1180,17 @@ MethodTableBuilder::bmtInterfaceEntry::CreateSlotTable(
...
+    MethodTable *interfaceMT = GetInterfaceType()->GetMethodTable();
+    SLOT_INDEX cSlots = interfaceMT->GetNumVirtuals();
+    if (interfaceMT->HasVirtualStaticMethods())
+    {
+        cSlots += interfaceMT->GetNumNonVirtualSlots();
+    }
     bmtInterfaceSlotImpl * pST = new (pStackingAllocator) bmtInterfaceSlotImpl[cSlots];
 
     MethodTable::MethodIterator it(GetInterfaceType()->GetMethodTable());
-    for (; it.IsValid(); it.Next())
```
Comment (davidwrighton) [inline src/coreclr/vm/methodtablebuilder.cpp]: I'd prefer to see the iterator as it is, and just check it.IsStatic and break for that case as well as the !Virtual case.

File: src/coreclr/vm/methodtablebuilder.cpp [hunk 3]
```diff
@@ -6314,24 +6320,28 @@ MethodTableBuilder::PlaceMethodImpls()
...
-                if (pCurDeclMethod->GetOwningType()->IsInterface())
-                {
-                    // Throws
-                    PlaceInterfaceDeclarationOnClass(
-                        pCurDeclMethod,
-                        pCurImplMethod);
-                }
-                else
+                // Don't place static virtual method overrides in the vtable
+                if (!IsMdStatic(pCurDeclMethod->GetDeclAttrs()))
```
Comment (davidwrighton) [inline src/coreclr/vm/methodtablebuilder.cpp]: This check should be immediately after the hDeclMethod.AsRTMethod() call. Also, it should be duplicated for the case above where the pCurDeclMethod is a MDMethod.

File: src/coreclr/vm/methodtablebuilder.cpp (hunk 2 above)
Comment (trylek) [reply src/coreclr/vm/methodtablebuilder.cpp]: Well, the problem is that the IsVirtual check is actually incorrect in this case - it just checks that the slot index is less then <code>GetNumVirtuals</code>.

File: src/coreclr/vm/methodtablebuilder.cpp (hunk 1 above)
Comment (trylek) [reply src/coreclr/vm/methodtablebuilder.cpp:5040]: Fixed in 4th commit, thanks for spotting!

File: src/coreclr/vm/methodtablebuilder.cpp (hunk 3 above)
Comment (trylek) [reply src/coreclr/vm/methodtablebuilder.cpp]: Fixed in 4th comment - I believe we can make the entire section after retrieval of <code>bmtMethodHandle hDeclMethod</code> conditional on non-static-ness.

File: src/coreclr/vm/methodtablebuilder.cpp (hunk 2 above)
Comment (trylek) [reply src/coreclr/vm/methodtablebuilder.cpp]: One other thing that seems somewhat concerning to me is that, when SVMs are present, we create the slot table for all non-virtual methods. We could easily place SVMs first followed by the other non-virtual methods and have another counter (<code>GetNumVirtualInstanceAndStaticMethods</code> or something) to account for the two subsets we need to generate the slot table for (it is used in <code>FindDeclMethodOnInterfaceEntry</code>).

Comment (trylek) [reply src/coreclr/vm/methodtablebuilder.cpp]: Fixed per our offline chat by reverting the changes to CreateSlotTable and short-circuiting ProcessMethodImpls in the presence of SVMs.

**Pull request merged**